library/psa_crypto.c

```diff
@@ -7442,15 +7460,27 @@ psa_status_t mbedtls_psa_crypto_configure_entropy_sources(
 
 void mbedtls_psa_crypto_free(void)
 {
+    /* Need to hold the mutex here to prevent this going ahead before
+     * psa_crypto_init() has completed, and to ensure integrity of
+     * global_data. */
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_mutex_lock(&mbedtls_threading_psa_globaldata_mutex);
+#endif /* defined(MBEDTLS_THREADING_C) */
```

Comment:
I am slightly icky about removing this for a couple of reasons.

1. I don't like saying that access to something is thread safe, then having thread un-safe accesses left to them, even though admittedly this function is not in itself threadsafe.
2. I think in the future, that we could potentially make this function thread safe, which means that it will have to look something like this, and I don't entirely want to lose that structure.
3. I don't think it does too much damage to expectations so long as the documentation (as it does) says "Don't use this in a threaded environment".

I do see your point, however, and I would like to see other opinions on this, before a final decision is made.

Reply:
`mbedtls_psa_crypto_free` is unlikely to need to be thread-safe. Its job is to clean up everything related to PSA, and it doesn't make sense to try and do that if other threads might need some of the PSA stuff.

Reply:
Synchronisation calls also act as memory barriers.  While it's clearly a problem if one thread shuts down the library while another thread is still using it, these things do happen.  If data on a weakly-ordered memory system isn't consistent, tracking down this mis-use is going to be harder for the user of the library.  So I think we should be thread-safe inside the library, even in shutdown code

Reply:
> So I think we should be thread-safe inside the library, even in shutdown code

In the future we might have that option, but at the moment we can't do that because even if we protect global data, the key store still won't be safe. (We can't make it so with the current constraints and design.)

Reply:
I still think we should protect the things that we can, i.e. the global data

Reply:
I'm of the opinion that we lose more than we gain by taking this out. We can't predict the future either way, but I agree with Tom that we should do what we can.